staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -816,6 +816,38 @@ func (c *Cacher) dispatchEvents() {
 	}
 }
 
+func setCachingObjects(event *watchCacheEvent, versioner storage.Versioner) {
+	if object, err := newCachingObject(event.Object); err == nil {
+		event.Object = object
+	} else {
+		klog.Errorf("couldn't create cachingObject from: %#v", event.Object)
+	}
+
+	// newCachingObject performs DeepCopy on the underlying object (which
+	// results in non-negligible number of memory allocations), thus we want
+	// to avoid doing that as much as possible.
+	// So we wrap event.Object into newCachingObject every time, but we only
+	// wrap event.PrevObject into newCachingObject when event.Object is nil.
+	// The reason for it is that event.PrevObject is used only to deliver
+	// DELETE watch event, so if event.Object is not nil it can be used only
+	// for watchers for which selector was satisfied for its previous version
+	// and is no longer satisfied for the current version. This case is rare
+	// enough, that the cost doesn't justify the gain for now.
+	if event.Object != nil {
```

Comment:
might be clearer to document in the returns why we don't care about wrapping PrevObject

if event.PrevObject == nil {
  // don't wrap nil PrevObject for create events
  return
}

if event.Object != nil {
  // don't wrap nil PrevObject for update events. we only encode those when ... and it's not worth the extra deep copies, etc
  return
}

Reply:
done